File: android/app/src/main/java/app/reyna/data/SmallTalk.kt
package app.reyna.data

/**
 * Whether something someone typed is a greeting rather than a question about
 * a document.
 *
 * Used only to suppress notes that make no sense as a reply to "hi". A reply
 * to a greeting was arriving with "833 files on this phone have not been sent
 * for reading yet, so they were not searched" stapled to it, which is a
 * perfectly good warning about a search and nonsense about a hello.
 *
 * Deliberately small and exact, and it errs towards saying no. Guessing wrong
 * in this direction only means a greeting keeps an irrelevant sentence;
 * guessing wrong the other way hides a real warning from somebody who was
 * genuinely looking for a file, and that warning exists to stop a false "I
 * have never seen that", which is the worst answer this app can give.
 */
object SmallTalk {

    private val PLEASANTRIES = setOf(
        "hi", "hii", "hiii", "hello", "helo", "hey", "heyy", "yo", "hola",
        "namaste", "greetings",
        "thanks", "thank", "thanku", "thankyou", "ty", "thx",
        "ok", "okay", "k", "kk", "cool", "nice", "great", "good", "fine",
        "morning", "afternoon", "evening", "night",
        "bye", "goodbye", "cya", "sup", "test", "testing", "you", "u", "there",
    )

    /**
     * True when every word is a pleasantry and there are at most three of
     * them, so "good morning" and "hey there" count but "good notes on
     * databases" does not.
     */
    fun matches(text: String): Boolean {
        val words = Words.of(text)
        if (words.isEmpty() || words.size > 3) return false
        return words.all { it in PLEASANTRIES }
    }
}

File: internal/nlp/smalltalk.go
package nlp

import (
	"strings"
	"unicode"
)

// pleasantries are words that carry no request.
var pleasantries = map[string]bool{
	"hi": true, "hii": true, "hiii": true, "hello": true, "helo": true,
	"hey": true, "heyy": true, "yo": true, "hola": true, "namaste": true,
	"greetings": true, "sup": true,
	"thanks": true, "thank": true, "thanku": true, "thankyou": true,
	"ty": true, "thx": true, "cheers": true,
	"ok": true, "okay": true, "k": true, "kk": true, "cool": true,
	"nice": true, "great": true, "good": true, "fine": true, "perfect": true,
	"morning": true, "afternoon": true, "evening": true, "night": true,
	"bye": true, "goodbye": true, "cya": true,
	"you": true, "u": true, "there": true, "a": true, "lot": true, "so": true, "much": true,
	"test": true, "testing": true,
	// Indic romanized & native pleasantries
	"dhanyawad": true, "dhanyawaad": true, "dhanyavad": true, "shukriya": true, "shukriyaa": true,
	"pranam": true, "pranaam": true, "vanakkam": true, "namaskaram": true, "alvida": true,
	"नमस्ते": true, "धन्यवाद": true, "शुक्रिया": true, "प्रणाम": true, "வணக்கம்": true, "నమస్కారం": true,
}

// IsSmallTalk reports whether a message is a greeting or an acknowledgement
// rather than a request for a document.
//
// Without this, "thanks" was parsed as a topic and searched for, and six
// unrelated documents came back with "I found 6 documents matching thanks, and
// they are about different things. Which one did you mean?". Saying thank you
// should not produce a disambiguation prompt.
//
// Deliberately narrow, and it errs towards saying no. A greeting misread as a
// question only wastes a search; a question misread as a greeting refuses to
// look for a document somebody actually wants. So every word has to be a
// pleasantry and there can be at most four of them: "thanks a lot" and "good
// morning" qualify, "good notes on databases" does not.
func IsSmallTalk(text string) bool {
	words := strings.FieldsFunc(strings.ToLower(text), func(r rune) bool {
		return !unicode.IsLetter(r) && !unicode.IsDigit(r) && !unicode.IsMark(r)
	})
	if len(words) == 0 || len(words) > 4 {
		return false
	}
	for _, w := range words {
		if !pleasantries[w] {
			return false
		}
	}
	return true
}
